File: core/esmf-aspect-meta-model-java/src/main/java/org/eclipse/esmf/metamodel/impl/DefaultCharacteristic.java
/*
 * Copyright (c) 2023 Robert Bosch Manufacturing Solutions GmbH
 *
 * See the AUTHORS file(s) distributed with this work for additional
 * information regarding authorship.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 */
package org.eclipse.esmf.metamodel.impl;

import java.util.Objects;
import java.util.Optional;
import java.util.StringJoiner;

import org.eclipse.esmf.metamodel.Characteristic;
import org.eclipse.esmf.metamodel.Type;
import org.eclipse.esmf.metamodel.loader.MetaModelBaseAttributes;
import org.eclipse.esmf.metamodel.visitor.AspectVisitor;

public class DefaultCharacteristic extends ModelElementImpl implements Characteristic {
   private final Optional<Type> dataType;

   public DefaultCharacteristic( final MetaModelBaseAttributes metaModelBaseAttributes, final Optional<Type> dataType ) {
      super( metaModelBaseAttributes );
      this.dataType = dataType;
   }

   /**
    * Defines the data type of all Properties which use this Characteristic.
    *
    * @return the dataType.
    */
   @Override
   public Optional<Type> getDataType() {
      return dataType;
   }

   /**
    * Accepts an Aspect visitor
    *
    * @param visitor The visitor to accept
    * @param <T> The result type of the traversal operation
    * @param <C> The context of the visitor traversal
    */
   @Override
   public <T, C> T accept( final AspectVisitor<T, C> visitor, final C context ) {
      return visitor.visitCharacteristic( this, context );
   }

   @Override
   public String toString() {
      return new StringJoiner( ", ", DefaultCharacteristic.class.getSimpleName() + "[", "]" )
            .add( "dataType=" + dataType )
            .toString();
   }
}

File: core/esmf-aspect-model-document-generators/src/test/resources/ts/snapshots/aspect-meta-model/characteristic/default-characteristic.ts
/*
 * Copyright (c) 2023 Robert Bosch Manufacturing Solutions GmbH
 *
 * See the AUTHORS file(s) distributed with this work for
 * additional information regarding authorship.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * SPDX-License-Identifier: MPL-2.0
 */

import {Type} from '../type';
import {Base, BaseMetaModelElement} from '../base';
import {ModelVisitor} from '../visitor/model-visitor';

export interface Characteristic extends BaseMetaModelElement {
    dataType?: Type;
}

export class DefaultCharacteristic extends Base implements Characteristic {
    constructor(metaModelVersion: string, aspectModelUrn: string, name: string, private _dataType?: Type) {
        super(metaModelVersion, aspectModelUrn, name);
    }

    public set dataType(value: Type | undefined) {
        this._dataType = value;
    }

    public get dataType(): Type | undefined {
        return this._dataType;
    }

    public accept<T, U>(visitor: ModelVisitor<T, U>, context: U): T {
        return visitor.visitCharacteristic(this, context);
    }

    /**
     * Get the effective type defined by possible nested characteristic definitions.
     */
    public static getEffectiveDataType(characteristic: Characteristic): Type | undefined {
        const characteristicAny: any = characteristic;
        if (characteristicAny.constructor.name === 'DefaultTrait') {
            return characteristicAny?.baseCharacteristic?.dataType;
        }
        if (characteristicAny.constructor.name === 'DefaultCollection' && characteristicAny.elementCharacteristic) {
            if (characteristicAny.elementCharacteristic.constructor.name === 'DefaultTrait') {
                return characteristicAny.elementCharacteristic.baseCharacteristic?.dataType;
            }
            return characteristicAny.elementCharacteristic.dataType || null;
        }
        return characteristicAny ? characteristicAny.dataType : null;
    }
}
